[search] Optimization of the new search.

ygorshenin opened the pull request.

mpimenov commented on search/search_query_factory.hpp:
```diff
@@ -25,11 +25,11 @@ class SearchQueryFactory
-    return make_unique<Query>(index, categories, suggests, infoGetter);
-#endif  // defined(USE_SEARCH_QUERY_V2)
+// #else
+//     return make_unique<Query>(index, categories, suggests, infoGetter);
```
Что тут происходит? Если V2 теперь безусловно, то почему ты закомментировал, а не стёр?

ygorshenin commented:
Following optimizations are made:
* implemented cache of nearby and matching streets
* implemented features filter (by viewport)
* optimized house numbers matching

ygorshenin commented:
@vng @mpimenov PTAL

vng commented on search/retrieval.hpp:
```diff
@@ -106,8 +106,12 @@ class Retrieval : public my::Cancellable
+
+  // Retrieves from the geometry index corresponding to |value| all features belonging to |rect|.
+  WARN_UNUSED_RESULT static unique_ptr<coding::CompressedBitVector> RetrieveGeometryFeatures(
+      MwmValue & value, my::Cancellable const & cancellable, m2::RectD const & rect, int scale);
```
Не совсем понятно зачем тут WARN_UNUSED_RESULT. Без возвращаемого значения звать функцию бессмысленно. По такой же логике, этот макрос можно лепить в любую функцию ...

ygorshenin replied:
Не обязательно в любую, но в те которые "тяжеловесные" желательно, чтобы некорректное использование отлавливалось компилятором.

vng replied:
Хз, мне кажется это уже синтаксический мусор и только мешает читать код. Реально, мы пол кода можем завалить этим макросом, по таким же соображениям. Я просто боюсь что с такими прецендентами, мы создадим свой макросовый язык и новичку вообще будет странно читать это все.

ygorshenin replied:
Могу рассказать историю как люди пролетели с полуфинала ACM ICPC из-за того что не воспользовались возвращаемым значением функции. У нас и так есть макросовый язык (UNIT_TEST, DISALLOW_COPY, WARN_UNUSED_RESULT, UNUSED_VALUE, ...) и не сказать что это плохо.

vng replied:
ну ладно

vng commented on search/search_query.cpp:
```diff
@@ -210,6 +207,9 @@ RankTable const * Query::RetrievalCallback::LoadTable(MwmSet::MwmId const & id)
 void Query::RetrievalCallback::UnloadTable(MwmSet::MwmId const & id) { m_rankTables.erase(id); }
 
+// static
+size_t const Query::kPreResultsCount;
```
Это все ок? Т.е. я к тому что наверное это определение не надо. Странно почему компилятор взял инициализацию в декларации переменной. Тогда это уже не declaration, a definition. Но с другой стороны мы имеем еще один definition в cpp.

ygorshenin replied:
Это ок. В хидере declaration. Это definition.

vng replied:
= 200 это разве declaration ? Я к тому что норм компилятору эта штука в cpp вообще не нужна.

ygorshenin replied:
= 200 это initialization.
Ты же не можешь завести глобальную переменную в хидере без extern - потому что она включится во все единицы трансляции и будет ODR violation. Так же и со статическими полями - где-то должна быть выделена ячейка памяти под нее на случай если кто-нибудь решит взять адрес. Поэтому определение переменной не может быть в хидере, так как компилятору не ясно где выделить память под переменную - хидер может быть как включен кем-либо так и не включен вообще. Однако, в хидере возможно объявление переменной + ее инициализация.

vng replied:
Так в том то и фишка, что под статическую интегральную константу память не нужна. Или я не прав? Она же константа.

ygorshenin replied:
Не нужна если не берется адрес. Пока в коде не берется, и если настаиваешь поправлю на constexpr.

ygorshenin replied:
constexpr не работает - undefined reference to Query::kPreResultsCount. Мб оставим как есть?

vng replied:
ок

vng commented on search/v2/features_filter.cpp:
```diff
@@ -0,0 +1,71 @@
+  else
+  {
+    m_featuresCache =
+        Retrieval::RetrieveGeometryFeatures(*m_value, m_cancellable, m_viewport, m_scale);
```
Не знаю, нужен ли тут scale. Я бы передавал максимальный scale всегда (GetUpperStyleScale()).
Чтобы кэшировались все фичи вне зависимости от их масштаба отрисовки.

ygorshenin replied:
Он и так равен этому значению, но протягивается из SearchQueryParams. Пусть лучше протягивается оттуда - если подадобится его поменять по всему поиску, достаточно будет поменять только в одном месте.

vng commented on search/v2/features_layer_matcher.hpp:
```diff
@@ -79,48 +84,41 @@ class FeaturesLayerMatcher
-          fn(child.m_sortedFeatures[i], sortedParentFeatures[j]);
+        if (rect.IsPointInside(childCenters[i]))
+          fn((*child.m_sortedFeatures)[i], (*parent.m_sortedFeatures)[j]);
       }
```
Лучше для locality получать покрытие и по нему получать id фич. И фильтровать потом по id от других токенов. Иначе нам надо получать childCenters[i] - а это опять надо делать вычитки.

vng commented on search/v2/features_layer_matcher.hpp:
```diff
@@ -144,34 +144,58 @@ class FeaturesLayerMatcher
+      // the query and feature's house numbers must be numbers, their
+      // first symbols must be the same.
+      string const houseNumber = feature.GetHouseNumber();
+      if (!queryLooksLikeHouseNumber || !feature::IsHouseNumber(houseNumber))
```
не понимаю, почему мы переменную queryLooksLikeHouseNumber проверяем в цикле?
Почему сразу не проверить и уйти там, где мы ее получили.

ygorshenin replied:
Потому что часть поискового запроса, на которую заматчились BUILDING, может и не выглядеть как номер дома, а быть, например, его названием.

vng commented on search/v2/geocoder.hpp:
```diff
@@ -56,32 +56,52 @@ class SearchModel;
+  {
+    Params();
+
+    m2::RectD m_viewport;
```
Не знаю как у тебя заложено дальше, но для ядра поиска в обзем случае у нас будет n-ректов первых приюлижений для наиболее вероятного результата.
- около меня
- около вьюпорта
- n-ректов полученных пре первычной обработке токенов как locality

ygorshenin replied:
Тогда это можно обобщить как вектор из m2::RectD?

ygorshenin replied:
Хотя я не очень понимаю зачем нам тут прямоугольники locality. Что такое locality - это определит сам geocoder. Но то что здесь должны быть координаты позиции пользователя я согласен.

vng replied:
Да, и если найдется локалити, то он даст первое приближение вьюпорта где искать. Грубо говоря, если есть слово Москва, то в одной из интерпретаций следует искать только в ее ректе.

ygorshenin replied:
Предлагаю это отдельным PR.

vng commented on search/v2/rank_table_cache.cpp:
```diff
@@ -0,0 +1,31 @@
+    return *it->second;
+  auto table = RankTable::Load(value.m_cont);
+  if (!table)
+    table.reset(new DummyRankTable());
```
А как работает DummyRank table? Он будет возврашать всегда одну и ту же константу или читать ывису нее осталост неизменным.

ygorshenin replied:
Он всегда возвращает 0.

ygorshenin commented:
PTAL

ygorshenin commented:
PTAL @vng @mpimenov

vng commented:
LGTM

mpimenov commented on search/v2/features_layer_matcher.hpp:
```diff
@@ -130,10 +128,12 @@ class FeaturesLayerMatcher
     vector<string> queryTokens;
     NormalizeHouseNumber(child.m_subQuery, queryTokens);
+    bool const queryLooksLikeHouseNumber =
+        feature::IsHouseNumber(child.m_subQuery) && !queryTokens.empty();
```
Мб поменять местами?

ygorshenin replied:
Done

mpimenov commented on search/v2/features_layer_matcher.hpp:
```diff
@@ -144,34 +144,58 @@ class FeaturesLayerMatcher
   ReverseGeocoder m_reverseGeocoder;
+
+  // Cache of streets in a feature's vicinity. All lists in the cache
+  // are ordered by a distance.
```
s/by a/by/

ygorshenin replied:
Done

mpimenov commented on search/v2/features_layer_matcher.hpp:
```diff
@@ -144,34 +144,58 @@ class FeaturesLayerMatcher
+  // Cache of correct streets for buildings. Current search algorithm
+  // supports only one street for a building, whereas buildings can be
+  // located on multiple streets.
+  unordered_map<uint32_t, uint32_t> m_matchingStreetsCache;
```
Вот кстати.
Сейчас один кэш на один layer_matcher. Может иметь смысл вынести его куда-нибудь выше?

ygorshenin replied:
Один LayerMatcher на одну mwm, поэтому нет смысла выносить выше. Т.е. понадобится скорее всего, но тогда придется делать двухуровневую таблицу, где первым ключом будет MwmId, а вторым фича. Или же один ключ - FeatureID. Но это лучше на потом.

mpimenov replied:
В этом случае возникает вопрос -- а когда эти кэши очищаются?

ygorshenin replied:
LayerMatcher сейчас пересоздается на каждую mwm, поэтому они очищаются в деструкторе LayerMatcher-а.

ygorshenin commented:
PTAL @mpimenov

mpimenov commented on search/v2/geocoder.cpp:
```diff
@@ -99,16 +137,26 @@ void Geocoder::Go(vector<FeatureID> & results)
+        PrepareRetrievalParams(i, i + 1);
+        m_partitions[i].FromFeatures(Retrieval::RetrieveAddressFeatures(
+                                         *m_value, *this /* cancellable */, m_retrievalParams),
+                                     *m_loader, m_model);
```
Форматирование съехало?

ygorshenin replied:
Нет, это clang-format.

mpimenov replied:
Оу.

mpimenov commented on search/v2/geocoder.cpp:
no line break?

ygorshenin commented on search/v2/geocoder.cpp:
Нет, это unix format.

mpimenov commented on search/v2/geocoder.cpp:
В чём именно тут unix format? Я предлагаю не разбивать один отрезок на две строки.

ygorshenin commented:
PTAL

mpimenov commented:
LGTM
